File: java/ex_046_say/Say.java
import java.util.ArrayList;
import java.util.List;

public class Say {
    private static final String[] LESS_THAN_20 = {
        "zero", "one", "two", "three", "four", "five", "six", "seven", "eight", "nine",
        "ten", "eleven", "twelve", "thirteen", "fourteen", "fifteen", "sixteen", "seventeen", "eighteen", "nineteen"
    };
    private static final String[] TENS = {
        "", "", "twenty", "thirty", "forty", "fifty", "sixty", "seventy", "eighty", "ninety"
    };
    private static final String[] SCALES = {"", "thousand", "million", "billion", "trillion"};

    public String say(long number) {
        if (number < 0 || number > 999_999_999_999L)
            throw new IllegalArgumentException("Number must be between 0 and 999,999,999,999.");
        if (number == 0)
            return "zero";
        List<String> parts = new ArrayList<>();
        int scaleIndex = 0;
        // Process the number in chunks of thousands
        while (number > 0) {
            long chunk = number % 1000;
            number /= 1000;
            if (chunk > 0)
                parts.add(convertChunk(chunk) + " " + SCALES[scaleIndex]);
            scaleIndex++;
        }
        // Reverse the parts to get the correct order
        StringBuilder result = new StringBuilder();
        for (int i = parts.size() - 1; i >= 0; i--) {
            if (i < parts.size() - 1)
                result.append(" ");
            result.append(parts.get(i).trim());
        }
        return result.toString().trim();
    }

    private String convertChunk(long chunk) {
        if (chunk == 0)  return "";
        if (chunk < 20)  return LESS_THAN_20[(int) chunk];
        if (chunk < 100) return TENS[(int) (chunk / 10)] + (chunk % 10 != 0 ? "-" + LESS_THAN_20[(int) (chunk % 10)] : "");
        return LESS_THAN_20[(int)(chunk / 100)] + " hundred" + (chunk % 100 != 0 ? " " + convertChunk(chunk % 100) : "");
    }
}

File: python/ex_059_say/say.py
one = ('','one','two','three','four','five','six','seven','eight','nine')
ten = ('ten','eleven','twelve','thirteen','fourteen','fifteen','sixteen','seventeen','eighteen','nineteen')
two = ('','','twenty','thirty','forty','fifty','sixty','seventy','eighty','ninety')
tho = ('','thousand','million','billion')
def say(number):
    if not 0 <= number < 10 ** 12: raise ValueError("input out of range")
    if not number: return 'zero'
    output = ''
    count = 0
    while number > 0:
        mil = number % 1000
        number //= 1000
        s = say_1(mil)
        if s: output = s + ' ' + tho[count] + ' ' + output
        count += 1
    return output.rstrip()
    
def say_1(number):
    if not number: return ''
    pos = -1
    unit = 0
    output = []
    while number > 0:
        pos += 1
        digit = number % 10
        number //= 10
        if pos % 3 == 0:
            output.insert(0, one[digit]) 
            unit = digit
        elif pos % 3 == 1:
            if digit > 1:
                output.insert(0, two[digit] + ('-' if unit else ''))
            elif digit:
                output.pop(0)
                output.insert(0, ten[unit])
        else:
            output.insert(0, one[digit] + ' hundred ' * (digit != 0))
    return ''.join(output).rstrip()